File: Pulkit_Chhabra/Basic_Data_Structures_Jan_2021/D1/2d_submatrix_sum.cpp
#include<bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
using namespace __gnu_pbds;
using namespace std;

#define ff              first
#define ss              second
#define int             long long
#define pb              push_back
#define mp              make_pair
#define mt              make_tuple
#define pii             pair<int,int>
#define vi              vector<int>
#define mii             map<int,int>
#define pqb             priority_queue<int>
#define pqs             priority_queue<int,vi,greater<int> >
#define setbits(x)      __builtin_popcountll(x)
#define mod             1000000007
#define inf             1e18
#define ps(x,y)         fixed<<setprecision(y)<<x
#define mk(arr,n,type)  type *arr=new type[n];
#define range(a,b)		substr(a,b-a+1)
#define w(x)            int x; cin>>x; while(x--)
#define trace(x) 		cerr<<#x<<": "<<x<<" "<<endl;
#define FIO             ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0)
mt19937                 rng(chrono::steady_clock::now().time_since_epoch().count());

typedef tree<int, null_type, less<int>, rb_tree_tag, tree_order_statistics_node_update> pbds;

int32_t main()
{
	FIO;
	int n, m; cin >> n >> m;
	vector<vi> arr(n + 1, vi(m + 1)), dp(n + 1, vi(m + 1));;

	for (int i = 1; i <= n; ++i)
		for (int j = 1; j <= m; ++j)
		{
			cin >> arr[i][j];
			dp[i][j] = arr[i][j] + dp[i - 1][j] + dp[i][j - 1] - dp[i - 1][j - 1];
		}

	w(q)
	{
		int i1, j1, i2, j2; cin >> i1 >> j1 >> i2 >> j2; //1-based

		int ans = (dp[i2][j2] - dp[i2][j1 - 1] - dp[i1 - 1][j2] + dp[i1 - 1][j1 - 1]);
		cout << ans << '\n';
	}


	return 0;
}

File: Pulkit_Chhabra/Basic_Data_Structures_Jan_2021/D1/2d_submatrix_sum.py
def main():
    n, m = map(int, input().split(' '))
    arr = [[0] * (m + 1)] * (n + 1)
    dp = [[0] * (m + 1)] * (n + 1)

    for i in range(1, n+1):
        arr[i] = list(map(int, input().split(' ')))
        for j in range(1, n+1):
            dp[i][j] = arr[i][j] + dp[i - 1][j] + \
                dp[i][j - 1] - dp[i - 1][j - 1]

    q = int(input())
    while q:
        q -= 1
        i1, j1, i2, j2 = map(int, input().split(' '))
        ans = (dp[i2][j2] - dp[i2][j1 - 1] -
               dp[i1 - 1][j2] + dp[i1 - 1][j1 - 1])
        print(ans)


if __name__ == '__main__':
    main()
